ansible/roles/test/tasks/pfc_wd/functional_test/functional_test.yml

```diff
@@ -0,0 +1,299 @@
+      vars:
+        ptf_test_name: PFC WD test
+        ptf_test_dir: ptftests
+        ptf_test_path: pfc_wd.PfcWdTest
+        ptf_platform: remote
+        ptf_test_params:
+          - testbed_type='{{testbed_type}}'
+          - router_mac='{{ansible_Ethernet0['macaddress']}}'
+          - queue_index='{{pfc_queue_index}}'
+          - pkt_count='10000'
+          - port_src='{{pfc_wd_test_port_id}}'
+          - ip_src='{{pfc_wd_rx_neighbor_addr}}'
+          - ip_dst='{{pfc_wd_test_neighbor_addr}}'
+          - wd_action='forward'
+        ptf_extra_options: "--relax --debug info --log-file /tmp/pfc_wd.PfcWdTest.{{ansible_date_time.iso8601}}.log "
+
+    - set_fact:
+        test_expect_file: "expect_pfc_wd_restore"
+
+    - name: Initialize loganalyzer
+      include: roles/test/files/tools/loganalyzer/loganalyzer_init.yml
+
+    - name: Wait for queue to recover from PFC storm
+      pause:
+        seconds: "{{(pfc_wd_penalty / 1000 * 2) | int}}"
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_analyze.yml
+
+    - name: Check if logs contain message that PFC WD restored from deadlock
+      include: roles/test/files/tools/loganalyzer/loganalyzer_end.yml
+
+  rescue:
+    - name: Remove all temporary files from local host.
+      file:
+        name: "{{ tests_location }}/functional_test/{{ item }}"
+        state: absent
+      with_items: "{{ config_files }}"
+      connection: local
+      become: false
+
+  always:
+    - name: Delete forward config from {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_del_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+    - name: Remove all the temporary files created by the test.
+      include: roles/test/tasks/deinit_config_test.yml
+
+# Verify timers
+- name: Initialize config test
+  include: roles/test/tasks/init_config_test.yml
+
+- block:
+    - name: Apply forward config to {{ pfc_wd_test_port }}.
+      vars:
+        config_file: pfc_wd_fwd_action.json
+        errors_expected: false
+        run_cleanup: false
+      include: roles/test/tasks/run_config_test.yml
+
+    - name: Enable routing from mgmt interface to localhost
+      sysctl:
+        name: net.ipv4.conf.eth0.route_localnet=1
+        value: 1
+        sysctl_set: yes
+
+      # Cannot use iptables module, because it does not support to_destination in 2.0
+    - name: Add iptables rule to forward syslog packets from mgmt interface to localhost
+      shell: "iptables -t nat -I PREROUTING -p udp -d {{ansible_eth0['ipv4']['address']}} --dport 514 -j DNAT --to-destination 127.0.0.1:514"
+
+    - set_fact:
+        detect_time_list: []
+        restore_time_list: []
+
+    - name: Calculate detection and restoration timings
+      include: roles/test/tasks/pfc_wd/functional_test/timer_test.yml
+      with_sequence: start=1 end=3
```

Comment:
3 times might be too small. And just want to understand what is the reason why we need to get the average instead of passing every time? What might be the reason of random error?

Reply:
3 times is the way usually used to verify such staff, why adding configurable parameter which will not be used?

Reply:
PFC generator is the weak link. I cannot guarantee that it is 100% reliable. But we can come up with other criteria if you have some suggestion